include/ginkgo/core/solver/solver_base.hpp

```diff
@@ -50,6 +51,126 @@ namespace gko {
+
+
+/**
+ * ApplyWithInitialGuess provides a way to give the input guess for apply
+ * function. All functionalities have the protected access specifier. It should
+ * only be used internally.
+ */
+class ApplyWithInitialGuess {
+protected:
+    friend class MultigridState;
+
+    /**
+     * Modify the input vector x by the guess
+     *
+     * @param b  the right hand side vectors
+     * @param x  the input vectors
+     * @param guess  the input guess
+     */
+    static void prepare_initial_guess(const LinOp* b, LinOp* x,
+                                      initial_guess_mode guess);
+
+    /**
+     * Applies a linear operator to a vector (or a sequence of vectors) with
+     * initial guess statement.
+     *
+     * Performs the operation x = op(b) with a initial guess statement, where op
+     * is this linear operator and initial guess statement will change the input
+     * vectors to fit the initial guess.
+     *
+     * @param b  the input vector(s) on which the operator is applied
+     * @param x  the output vector(s) where the result is stored
+     * @param guess  the input guess to handle the input vector(s)
+     *
+     * @return this
+     */
+    virtual void apply_with_initial_guess(const LinOp* b, LinOp* x,
+                                          initial_guess_mode guess) const
+    {
+        prepare_initial_guess(b, x, guess);
+        this->apply_impl(b, x);
+    }
+
+    /**
+     * Performs the operation x = alpha * op(b) + beta * x with a initial guess
+     * statement, where op is this linear operator and initial guess statement
+     * will change the input vectors to fit the initial guess
+     *
+     * @param alpha  scaling of the result of op(b)
+     * @param b  vector(s) on which the operator is applied
+     * @param beta  scaling of the input x
+     * @param x  output vector(s)
+     * @param guess  the input guess
+     *
+     * @return this
+     */
+    virtual void apply_with_initial_guess(const LinOp* alpha, const LinOp* b,
+                                          const LinOp* beta, LinOp* x,
+                                          initial_guess_mode guess) const
+    {
+        prepare_initial_guess(b, x, guess);
+        this->apply_impl(alpha, b, beta, x);
+    }
+
+    /**
+     * Get the default initial guess
+     *
+     * @return default initial guess
+     */
+    initial_guess_mode get_default_initial_guess() const { return guess_; }
+
+    /**
+     * ApplyWithInitialGuess constructor.
+     *
+     * @param guess  the input guess whose default is
+     * initial_guess_mode::provided
+     */
+    explicit ApplyWithInitialGuess(
+        initial_guess_mode guess = initial_guess_mode::provided)
+        : guess_(guess)
+    {}
+
+    /**
+     * set the default initial guess
+     *
+     * @param guess  the initial guess
+     */
+    void set_default_initial_guess(initial_guess_mode guess) { guess_ = guess; }
+
+    // override this at the same time when overridden
```

Comment:
Does this actually interact with the `LinOp::apply_impl` or does it just lead to ambiguity? Without any subclass relation to `LinOp`, I'm not sure this does what you intend it to do

Reply:
because all solver inherit this class, it provide the default behavior.
change the input x according to the initial_guess_mode.
By doing so, we do not need to reimplement all apply_with_initial_guess when there's no benefit, or we have no time to doing that. I do not want to do that in this pr. Otherwise, I would change the whole code interface three/four times before.

Reply:
I think it is. https://godbolt.org/z/6YrhP6nKz

Reply:
In any case, I think this is a rather confusing design, seeing as there is no relation between ApplyWithInitialGuess and LinOp - we are somewhere between a pure interface and an implementation here, and I'd rather see it only be one of those two things. So my suggestion is: Make all virtual functions in here pure, and instead implement the default behavior in a mixin, which can then rely on the base class being derived from LinOp.